### DISCUSSION TITLE
why is there no cookies object on context.res in getServerSideProps?

### DISCUSSION THREAD

**Discussion by patrick99e99** [vercel/next.js#70765] [Help] [answered]
### Summary

I am trying to wrap my head around why it seems the apis for requests and responses seem to be different all over the place..

In a middleware file, I can do `nextResponse.cookies.set('foo', 'bar')` ... but if I try to do the same in the server side props function with `context.res.cookies.set` it says: ` Property 'cookies' does not exist on type 'ServerResponse'.` 

It seems that in order to set a cookie in the server side props I have to `context.res.setHeader('set-cookie', ['some_cookie=some_val'])`

And I just am wondering why this is...?  Why isn't there a consistent interface for a response object?

**Comment by patrick99e99** [comment]
It seems that if I do:

```
    const res = context.res as Response // 'express' response type
    res.cookie('foo', 'bar')
```

then this will work..  But I am still wondering, why do I need to do that?  Why doesn't nextjs' interface for `context.res` know that cookie exists on it?  And `res.cookie('foo', 'bar')` is different than middleware where I was doing `res.cookies.set('foo', bar')`..  I just am confused why this is different everywhere.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The problem here is that GSSP (getServerSideProps), uses `IncomingMessage`, https://nodejs.org/api/http.html#class-httpincomingmessage, and `ServerResponse`, https://nodejs.org/api/http.html#http_class_http_serverresponse, which are Node.js constructs, whereas, other parts of Next.js aligned behind, a more recent standardisation, so called, Web APIs. 

In Web APIs, one can find the Fetch API family, which includes, for example, the `fetch` function, the `Request`, and `Response` objects. 

So that's the core reason, as to why, GSSP, GSP, and other Page Router elements, seem to have a different surface API for things such as headers, cookies, etc. 

To reiterate, it is because, these came way before the standardisation of other Web APIs into runtimes outside of a browser, that is to say, they originated in the Node.js runtime, which as you may now, only recently (almost 1.5 years now), started to support things such as fetch.

**Reply by icyJoseph** [reply]
@patrick99e99 please don't forget to mark as answered!
